cmake/modules/RootMacros.cmake

```diff
@@ -836,35 +954,66 @@ function(ROOT_LINKER_LIBRARY library)
 endfunction()
 
 #---------------------------------------------------------------------------------------------------
-#---ROOT_ADD_INCLUDE_DIRECTORIES( library )
+#---ROOT_CONFIGURE_LIBRARY_INCLUDES( library DIRS incl1 incl2 ... GDIRS incl1 incl2 ... V7DIRS incl1 incl2 ...)
+#---------------------------------------------------------------------------------------------------
+function(ROOT_CONFIGURE_LIBRARY_INCLUDES library)
+  CMAKE_PARSE_ARGUMENTS(ARG "" "" "DIRS;GDIRS;V7DIRS" ${ARGN})
+  set(lst)
+  foreach(glbl ${ARG_GDIRS})
+     list(APPEND lst ${glbl})
+  endforeach()
+  foreach(src ${ARG_DIRS})
+     list(APPEND lst ${CMAKE_SOURCE_DIR}/${src})
+  endforeach()
+  if(root7)
+     foreach(src ${ARG_V7DIRS})
+        list(APPEND lst ${CMAKE_SOURCE_DIR}/${src})
+     endforeach()
+  endif(root7)
+
+  SET(root_incdirs_${library} ${lst} CACHE STRING "includes for ${library}" FORCE)
+endfunction()
+
+#---------------------------------------------------------------------------------------------------
+#---ROOT_ADD_INCLUDE_DIRECTORIES( library DEPENDENCIES dep1 dep2 ... INCLUDES incl1 incl2 ...)
 #---------------------------------------------------------------------------------------------------
 function(ROOT_ADD_INCLUDE_DIRECTORIES library)
+  CMAKE_PARSE_ARGUMENTS(ARG "" "" "DEPENDENCIES;INCLUDES" ${ARGN})
   if(PROJECT_NAME STREQUAL "ROOT")
+      list(APPEND used_libs Core)
+      list(APPEND used_libs ${ARG_DEPENDENCIES})
+      list(APPEND used_libs ${library})
+      list(REMOVE_DUPLICATES used_libs)
 
-    if(IS_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/inc)
-      target_include_directories(${library}
-        PRIVATE
-          ${CMAKE_CURRENT_SOURCE_DIR}/inc
-        INTERFACE
-          $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/inc>
-      )
-    endif()
+      set(fulllst)
 
-    if(root7 AND IS_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/v7/inc)
-      target_include_directories(${library}
-        PRIVATE
-          ${CMAKE_CURRENT_SOURCE_DIR}/v7/inc
-        INTERFACE
-          $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/v7/inc>
-      )
-    endif()
+      foreach(lib ${used_libs})
+         foreach(incl ${root_incdirs_${lib}})
+            list(APPEND fulllst ${incl})
+         endforeach()
+      endforeach()
 
-    if(IS_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/res)
-      target_include_directories(${library}
-        PRIVATE
-          ${CMAKE_CURRENT_SOURCE_DIR}/res
-      )
-    endif()
+      foreach(incl ${ARG_INCLUDES})
+        list(APPEND fulllst ${incl})
+      endforeach()
+
+      if(IS_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/inc)
+         list(APPEND fulllst ${CMAKE_CURRENT_SOURCE_DIR}/inc)
+       endif()
+
+      if(root7 AND (IS_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/v7/inc))
```

Comment:
See my comment above - we should try to reduce the amount of duplication of "and these are the default include dirs".

Reply:
It is old code, I did not modify it.
To remove it, I need in each `ROOT_LINKER_LIBRARY`, `ROOT_MODULE_LIBRARY`, `ROOT_OBJECT_LIBRARY` provide proper include paths, taking into account `if(root7)` condition.

Reply:
Include paths should be inherited by linking to targets that provide them in their build interface. Also, the directories in `core/*/inc` should be propagated from linking to the `Core` target, not added by hand to every target like is being done.